### PR TITLE
Add settings for doing secure auth LDAPS/AD

### PR DESCRIPTION
**Pull request by wrender**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no|yes
| API breaks?     | no|yes
| Deprecations?   | no|yes
| Related tickets | fixes #X, partially #Y, mentioned in #Z
| License         | Apache 2.0

### What's in this PR?
Adding settings required to do secure LDAP with nifi

### Why?
Missing settings for doing secure LDAP

### Checklist

- [ ] Implementation tested
- [ ] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [ ] Logging code meets the guideline
- [ ] User guide and development docs updated (if needed)
- [ ] Append changelog with changes

### To Do

- [ ] If the PR is not complete but you want to discuss the approach, list what remains to be done here

### CODE REVIEW

File: pkg/resources/templates/config/login_identity_providers.go [hunk 1]
```diff
@@ -73,17 +73,18 @@ var LoginIdentityProvidersTemplate = `<?xml version="1.0" encoding="UTF-8" stand
...
+        <property name="TLS - Truststore">{{.LdapConfiguration.TLSTruststore}}</property>
+        <property name="TLS - Truststore Password">{{.LdapConfiguration.TLSTruststorePassword}}</property>
+        <property name="TLS - Truststore Type">{{.LdapConfiguration.TLSTruststoreType}}</property>
+        <property name="TLS - Client Auth">{{.LdapConfiguration.ClientAuth}}</property>
+        <property name="TLS - Protocol">{{.LdapConfiguration.Protocol}}</property>
+        <property name="TLS - Shutdown Gracefully">{{.LdapConfiguration.ShutdownGracefully}}</property>
         
-        <property name="Referral Strategy">FOLLOW</property>
+        <property name="Referral Strategy">{{.LdapConfiguration.ReferralStrategy}}</property>
+        <property name="Referral Strategy">{{or .LdapConfiguration.ReferralStrategy "FOLLOW"}}</property>
```
Comment (mh013370) [inline pkg/resources/templates/config/login_identity_providers.go:87]:
```suggestion
        <property name="Referral Strategy">{{or .LdapConfiguration.ReferralStrategy "FOLLOW"}}</property>
```

**Comment by mh013370** [comment]
Would you mind updating the LDAP documentation here for the new configurable fields?

https://github.com/konpyutaika/nifikop/blob/master/site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md?plain=1#L189-L196

File: api/v1/nificluster_types.go [hunk 2]
```diff
@@ -580,7 +580,19 @@ type LdapConfiguration struct {
 	// The default functionality if this property is missing is USE_DN in order to retain backward compatibility.
 	// USE_DN will use the full DN of the user entry if possible.
 	// USE_USERNAME will use the username the user logged in with.
+	TLSKeystore string `json:"tlsKeystore,omitempty"`
```
[the remark this answers is not in the document]
Comment (wrender) [reply api/v1/nificluster_types.go:581]: Ok. Done

**Comment by wrender** [comment]
> Would you mind updating the LDAP documentation here for the new configurable fields?
> 
> https://github.com/konpyutaika/nifikop/blob/master/site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md?plain=1#L189-L196

Ok. Done

File: api/v1/nificluster_types.go [hunk 3]
```diff
@@ -576,11 +576,34 @@ type LdapConfiguration struct {
 	ManagerDn string `json:"managerDn,omitempty"`
 	// The password of the manager that is used to bind to the LDAP server to search for users.
 	ManagerPassword string `json:"managerPassword,omitempty"`
+    // Path to the Keystore that is used when connecting to LDAP using LDAPS or START_TLS.
+	// The TLS Keystore settings are optional and only used if your LDAP/AD server needs mutual TLS.
+	TLSKeystore string `json:"tlsKeystore,omitempty"`
```
Comment (juldrixx) [inline api/v1/nificluster_types.go:581]:
```suggestion
	// Path to the Keystore that is used when connecting to LDAP using LDAPS or START_TLS.
	// The TLS Keystore settings are optional and only used if your LDAP/AD server needs mutual TLS.
	TLSKeystore string `json:"tlsKeystore,omitempty"`
```

**Review by juldrixx — CHANGES_REQUESTED** [review]

**Comment by wrender** [comment]
For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container.  Am I missing something?  For example:
```
ldapConfiguration
  enabled: true
  tlsTruststore:  /some/path
```
Then if I exec into the container, and cat conf/login-identity-providers.xml  , The value for that line item is empty.

File: site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md [hunk 4]
```diff
@@ -188,12 +188,26 @@ spec:
...
+| tlsTruststoreType       | string | Type of the Truststore that is used when connecting to LDAP using LDAPS or START_TLS (i.e. JKS or PKCS12).                                 | No       | ""          |
+| clientAuth              | string | Client authentication policy when connecting to LDAP using LDAPS or START_TLS. Possible values are REQUIRED, WANT, NONE.                   | No       | ""          |
+| protocol                | string | Protocol to use when connecting to LDAP using LDAPS or START_TLS. (i.e. TLS, TLSv1.1, TLSv1.2, etc).                                       | No       | ""          |
+| shutdownGracefully      | string | Specifies whether the TLS should be shut down gracefully before the target context is closed. Defaults to false.                           | No       | ""          |
+| referralStrategy        | string | Strategy for handling referrals. Possible values are FOLLOW, IGNORE, THROW.                                                                | No       | ""          |
```
Comment (mh013370) [inline site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md:209]:
```suggestion
| referralStrategy        | string | Strategy for handling referrals. Possible values are FOLLOW, IGNORE, THROW.                                                                | No       | FOLLOW          |
```

**Comment by mh013370** [comment]
Just one minor doc suggestion. otherwise it looks good to me

**Comment by wrender** [comment]
> > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > ```
> > ldapConfiguration
> >   enabled: true
> >   tlsTruststore:  /some/path
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> 
> Just double checking you've applied the new CRDs here and the updated operator?

I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?

**Comment by mh013370** [comment]
> > > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > > ```
> > > ldapConfiguration
> > >   enabled: true
> > >   tlsTruststore:  /some/path
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> > 
> > 
> > Just double checking you've applied the new CRDs here and the updated operator?
> 
> I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?

There's a thread in slack where folks talk about that: https://konpytika.slack.com/archives/C0362VBRM24/p1702911625342399

TLDR is that the helm client doesn't do it for you, but tools like ArgoCD/Flux will do it for you. 

https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

**Comment by wrender** [comment]
> > > > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > > > ```
> > > > ldapConfiguration
> > > >   enabled: true
> > > >   tlsTruststore:  /some/path
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> > > 
> > > 
> > > Just double checking you've applied the new CRDs here and the updated operator?
> > 
> > 
> > I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?
> 
> There's a thread in slack where folks talk about that: https://konpytika.slack.com/archives/C0362VBRM24/p1702911625342399
> 
> TLDR is that the helm client doesn't do it for you, but tools like ArgoCD/Flux will do it for you.
> 
> https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations

I manually deleted the CRDs, and confirmed they are updated after re-installing the nifikop helm chart. But still, when I deploy a nifi cluster it doesn't seem to add the ldap settings. Seems like it must be something else then.

**Comment by juldrixx** [comment]
> > > > > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > > > > ```
> > > > > ldapConfiguration
> > > > >   enabled: true
> > > > >   tlsTruststore:  /some/path
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> > > > 
> > > > 
> > > > Just double checking you've applied the new CRDs here and the updated operator?
> > > 
> > > 
> > > I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?
> > 
> > 
> > There's a thread in slack where folks talk about that: https://konpytika.slack.com/archives/C0362VBRM24/p1702911625342399
> > TLDR is that the helm client doesn't do it for you, but tools like ArgoCD/Flux will do it for you.
> > https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations
> 
> I manually deleted the CRDs, and confirmed they are updated after re-installing the nifikop helm chart. But still, when I deploy a nifi cluster it doesn't seem to add the ldap settings. Seems like it must be something else then.

Did you build an image of the operator with your change? Or are you using the `1.7.0`? Or are you running the code locally?

**Comment by wrender** [comment]
> > > > > > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > > > > > ```
> > > > > > ldapConfiguration
> > > > > >   enabled: true
> > > > > >   tlsTruststore:  /some/path
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> > > > > 
> > > > > 
> > > > > Just double checking you've applied the new CRDs here and the updated operator?
> > > > 
> > > > 
> > > > I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?
> > > 
> > > 
> > > There's a thread in slack where folks talk about that: https://konpytika.slack.com/archives/C0362VBRM24/p1702911625342399
> > > TLDR is that the helm client doesn't do it for you, but tools like ArgoCD/Flux will do it for you.
> > > https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations
> > 
> > 
> > I manually deleted the CRDs, and confirmed they are updated after re-installing the nifikop helm chart. But still, when I deploy a nifi cluster it doesn't seem to add the ldap settings. Seems like it must be something else then.
> 
> Did you build an image of the operator with your change? Or are you using the `1.7.0`? Or are you running the code locally?

Running the code locally. I git cloned my fork, that has the login providers changes for ldap, and then do a helm install of nifikop from that local folder.

**Comment by juldrixx** [comment]
> > > > > > > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > > > > > > ```
> > > > > > > ldapConfiguration
> > > > > > >   enabled: true
> > > > > > >   tlsTruststore:  /some/path
> > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > >     
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >     
> > > > > > >   
> > > > > > > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> > > > > > 
> > > > > > 
> > > > > > Just double checking you've applied the new CRDs here and the updated operator?
> > > > > 
> > > > > 
> > > > > I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?
> > > > 
> > > > 
> > > > There's a thread in slack where folks talk about that: https://konpytika.slack.com/archives/C0362VBRM24/p1702911625342399
> > > > TLDR is that the helm client doesn't do it for you, but tools like ArgoCD/Flux will do it for you.
> > > > https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations
> > > 
> > > 
> > > I manually deleted the CRDs, and confirmed they are updated after re-installing the nifikop helm chart. But still, when I deploy a nifi cluster it doesn't seem to add the ldap settings. Seems like it must be something else then.
> > 
> > 
> > Did you build an image of the operator with your change? Or are you using the `1.7.0`? Or are you running the code locally?
> 
> Running the code locally. I git cloned my fork, that has the login providers changes for ldap, and then do a helm install of nifikop from that local folder.

If you didn't build an image of your code  to use in your deployment, it won't work. It will just deploy the latest release of the operator. If you want to run the code locally, you need to use a tool like [telepresence](https://www.telepresence.io/).

**Comment by juldrixx** [comment]
> > > > > > > > > For some reason when I define these new settings in a cluster crd yaml file, and then deploy a cluster they don't take affect in the container. Am I missing something? For example:
> > > > > > > > > ```
> > > > > > > > > ldapConfiguration
> > > > > > > > >   enabled: true
> > > > > > > > >   tlsTruststore:  /some/path
> > > > > > > > > ```
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > >     
> > > > > > > > >       
> > > > > > > > >     
> > > > > > > > > 
> > > > > > > > >       
> > > > > > > > >     
> > > > > > > > > 
> > > > > > > > >     
> > > > > > > > >   
> > > > > > > > > Then if I exec into the container, and cat conf/login-identity-providers.xml , The value for that line item is empty.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Just double checking you've applied the new CRDs here and the updated operator?
> > > > > > > 
> > > > > > > 
> > > > > > > I deleted the CRDs and re-created them with the helm chart. Maybe I'm not updating the operator correctly? How would I update that?
> > > > > > 
> > > > > > 
> > > > > > There's a thread in slack where folks talk about that: https://konpytika.slack.com/archives/C0362VBRM24/p1702911625342399
> > > > > > TLDR is that the helm client doesn't do it for you, but tools like ArgoCD/Flux will do it for you.
> > > > > > https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations
> > > > > 
> > > > > 
> > > > > I manually deleted the CRDs, and confirmed they are updated after re-installing the nifikop helm chart. But still, when I deploy a nifi cluster it doesn't seem to add the ldap settings. Seems like it must be something else then.
> > > > 
> > > > 
> > > > Did you build an image of the operator with your change? Or are you using the `1.7.0`? Or are you running the code locally?
> > > 
> > > 
> > > Running the code locally. I git cloned my fork, that has the login providers changes for ldap, and then do a helm install of nifikop from that local folder.
> > 
> > 
> > If you didn't build an image of your code to use in your deployment, it won't work. It will just deploy the latest release of the operator. If you want to run the code locally, you need to use a tool like [telepresence](https://www.telepresence.io/).
> 
> Ok. Thanks for the information @juldrixx . I'm new to operators in Kubernetes so I will have to spend some time learning this. Is there any documentation on how I would go about the building of a custom image to test with the deployment? I don't know I want to introduce yet another tool like telepresence.

You can find it [here](https://konpyutaika.github.io/nifikop/docs/6_contributing/1_developer_guide) but it doesn't mention telepresence.

**Review by mh013370 — APPROVED** [review]

**Comment by mh013370** [comment]
LGTM

**Review by juldrixx — APPROVED** [review]
